packages/yew/src/html/component/lifecycle.rs

```diff
@@ -1,111 +1,81 @@
 //! Component lifecycle module
 
-use super::{AnyScope, BaseComponent, Scope};
-use crate::html::RenderError;
+use super::scope::{AnyScope, Scope};
+use crate::dom_bundle::ComponentRenderState;
+use crate::html::{BaseComponent, RenderError};
 use crate::scheduler::{self, Runnable, Shared};
 use crate::suspense::{Suspense, Suspension};
-use crate::virtual_dom::{VDiff, VNode};
 use crate::Callback;
 use crate::{Context, NodeRef};
-#[cfg(feature = "ssr")]
-use futures::channel::oneshot;
 use std::rc::Rc;
 use web_sys::Element;
 
-pub(crate) struct ComponentState<COMP: BaseComponent> {
-    pub(crate) component: Box<COMP>,
-    pub(crate) root_node: VNode,
+pub struct ComponentState<COMP: BaseComponent> {
+    pub(super) component: Box<COMP>,
+    pub(super) render_state: ComponentRenderState,
 
     context: Context<COMP>,
-
-    /// When a component has no parent, it means that it should not be rendered.
-    parent: Option<Element>,
-
-    next_sibling: NodeRef,
     node_ref: NodeRef,
     has_rendered: bool,
 
     suspension: Option<Suspension>,
 
-    #[cfg(feature = "ssr")]
-    html_sender: Option<oneshot::Sender<VNode>>,
-
     // Used for debug logging
     #[cfg(debug_assertions)]
-    pub(crate) vcomp_id: u64,
+    vcomp_id: u64,
 }
 
 impl<COMP: BaseComponent> ComponentState<COMP> {
-    pub(crate) fn new(
-        parent: Option<Element>,
-        next_sibling: NodeRef,
-        root_node: VNode,
+    fn new(
+        initial_render_state: ComponentRenderState,
         node_ref: NodeRef,
         scope: Scope<COMP>,
         props: Rc<COMP::Properties>,
-        #[cfg(feature = "ssr")] html_sender: Option<oneshot::Sender<VNode>>,
     ) -> Self {
         #[cfg(debug_assertions)]
-        let vcomp_id = {
-            use super::Scoped;
-
-            scope.to_any().vcomp_id
-        };
+        let vcomp_id = { scope.vcomp_id };
         let context = Context { scope, props };
 
         let component = Box::new(COMP::create(&context));
         Self {
             component,
-            root_node,
+            render_state: initial_render_state,
             context,
-            parent,
-            next_sibling,
             node_ref,
             suspension: None,
             has_rendered: false,
 
-            #[cfg(feature = "ssr")]
-            html_sender,
-
             #[cfg(debug_assertions)]
             vcomp_id,
         }
     }
 }
 
-pub(crate) struct CreateRunner<COMP: BaseComponent> {
-    pub(crate) parent: Option<Element>,
-    pub(crate) next_sibling: NodeRef,
-    pub(crate) placeholder: VNode,
-    pub(crate) node_ref: NodeRef,
-    pub(crate) props: Rc<COMP::Properties>,
-    pub(crate) scope: Scope<COMP>,
-    #[cfg(feature = "ssr")]
-    pub(crate) html_sender: Option<oneshot::Sender<VNode>>,
+pub struct CreateRunner<COMP: BaseComponent> {
+    pub initial_render_state: ComponentRenderState,
+    pub node_ref: NodeRef,
+    pub props: Rc<COMP::Properties>,
```

Comment:
Just wondering, can we use `*const COMP::Properties` here?

Reply:
Is it worth the unsafe here if Rc is capable of being cloned tens of millions of times per second? (This data is based on a previous benchmark of Arc, Rc should be more efficient.)

In addition, Properties needs to be heap-allocated here (boxing) in order to live ‘static which is probably the most expensive operation and unavoidable regardless of Rc or raw pointer.

In addition, modern computers also capable of boxing millions of () per second.

Reply:
What is possible (and I think worth doing) is removing the Boxes in `scheduler` and storing and `Rc<dyn smth>` to the `ComponentState<COMP>` instead.